src/Microsoft.Azure.WebJobs.Host/Dispatch/SharedQueueHandler.cs

```diff
@@ -0,0 +1,261 @@
+using Microsoft.Azure.WebJobs.Host.Queues.Listeners;
+using Microsoft.Azure.WebJobs.Host.Storage.Queue;
+using Microsoft.Azure.WebJobs.Host.Timers;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.WebJobs.Host.Dispatch
+{
+    internal class SharedQueueHandler
+    {
+        private readonly IStorageAccountProvider _accountProvider;
+        private readonly IHostIdProvider _hostIdProvider;
+        private readonly IWebJobsExceptionHandler _exceptionHandler;
+        private readonly ILoggerFactory _loggerFactory;
+        private readonly TraceWriter _trace;
+        private readonly IQueueConfiguration _queueConfiguration;
+        private readonly ISharedContextProvider _sharedContextProvider;
+        private readonly IContextSetter<IMessageEnqueuedWatcher> _messageEnqueuedWatcherSetter;
+
+        private SharedQueueExecutor _triggerExecutor;
+        private State _state;
+        private QueueListener _sharedQueuelistener;
+        private SharedQueueWriter _sharedQueueWriter;
+
+        internal SharedQueueHandler(IStorageAccountProvider accountProvider,
+                           IHostIdProvider hostIdProvider,
+                           IWebJobsExceptionHandler exceptionHandler,
+                           TraceWriter trace,
+                           ILoggerFactory loggerFactory,
+                           IQueueConfiguration queueConfiguration,
+                           ISharedContextProvider sharedContextProvider,
+                           IContextSetter<IMessageEnqueuedWatcher> messageEnqueuedWatcherSetter)
+        {
+            _accountProvider = accountProvider;
+            _hostIdProvider = hostIdProvider;
+            _exceptionHandler = exceptionHandler;
+            _trace = trace;
+            _loggerFactory = loggerFactory;
+            _queueConfiguration = queueConfiguration;
+            _sharedContextProvider = sharedContextProvider;
+            _messageEnqueuedWatcherSetter = messageEnqueuedWatcherSetter;
+            _state = State.Created;
+        }
+
+        // this is used to prevent illegal call sequence
+        private enum State
+        {
+            Created,
+            Initialized,
+            Started,
+            Stopped,
+        }
+
+        internal Exception InitializationEx { get; private set; }
+
+        internal async Task StopQueueAsync(CancellationToken cancellationToken)
+        {
+            if (_state != State.Started)
+            {
+                throw new InvalidOperationException(ErrorMessage(State.Started, _state));
+            }
+            // the count value does not change once SharedQueue is started, see "RegisterHandler()"
+            if (_triggerExecutor.HasMessageHandlers())
+            {
+                await _sharedQueuelistener.StopAsync(cancellationToken);
+            }
+            // if there's no messageHandlers registed, stopQueue is a NOOP
+            _state++;
```

Comment:
>_state++; [](start = 12, length = 9)

Can you avoid ++ to trasntiion between statemachine states. That's too brittle on enum ordering. Can you explicitly set it ("_state = State.Stopped")